### PR TITLE
Add `"toggle"` select mode and make it the default in `TapTool`

### LINKED ISSUE #13831: Tap tool default mode should select+unselect, but not append
Original ref: https://discourse.bokeh.org/t/taptool-on-a-pie-chart-reopens-all-previous-taps/11448/6

In Bokeh 3.4.x the tap tool seems to be in permanent "append" mode, i.e. it always does what I would only expect to see if `shift` was added to a tap. 

This can be seen in the [`color_scatter` example in in the gallery](https://docs.bokeh.org/en/3.4.1/docs/examples/basic/scatters/color_scatter.html). Here is the result of just randomly clicking many places with no shift or other modifier added:

Screenshot 2024-04-17 at 10 35 46

This does not occur with the 3.3.x version of the gallery example -- each new tap replaces the previous selection. 

cc @bokeh/dev I have triaged `discussion` for now but I believe this is a bug/regression (and merits a 3.4.2 release).

### PR DESCRIPTION
**Pull request by mattpap**
Adds support for new `"toggle"` selection mode, which for `TapTool`, where it makes most sense, allows to select and deselect a datapoint, but not add to a selection. So it works almost like the existing and previously default `"replace"` mode, but allows deselection. This preserves original functionality of `TapTool` while implementing the requested deselection feature.

Currently this is enabled for all select tools, primarily because we don't have appropriate property overrides mechanisms in bokeh and bokehjs (I implemented this in bokeh in another PR) and that the selection logic doesn't care about tools, only about modes and all need to be implemented.

Using [`color_scatter.py`](https://github.com/bokeh/bokeh/blob/branch-3.5/examples/basic/scatters/color_scatter.py):

[Screencast from 29.05.2024 11:48:47.webm](https://github.com/bokeh/bokeh/assets/27475/445e73b0-f66d-4ac9-8f5a-330cfc0cba67)

fixes #13831

### CODE REVIEW

**Comment by mattpap** [comment]
I separated the generic (and used by `TapTool`) `SelectionMode` from `RegionSelectionMode` (used by `BoxSelectTool`, etc.).

File: bokehjs/src/lib/core/util/arrayable.ts [hunk 1]
```diff
@@ -35,7 +38,20 @@ export function copy<T>(array: Arrayable<T>): Arrayable<T> {
...
+export function splice<T>(array: T[], start: number, k?: number, ...items: T[]): T[]
+export function splice<T>(array: Arrayable<T>, start: number, k?: number, ...items: T[]): Arrayable<T>
+
 export function splice<T>(array: Arrayable<T>, start: number, k?: number, ...items: T[]): Arrayable<T> {
+  if (Array.isArray(array)) {
+    const result = copy(array)
+    if (k === undefined) {
+      result.splice(start)
+    } else {
+      result.splice(start, k, ...items)
```
Comment (hoxbro) [inline bokehjs/src/lib/core/util/arrayable.ts:50]: Are there any reasons not to use toSpliced here?

File: bokehjs/src/lib/core/util/arrayable.ts [hunk 2]
```diff
@@ -35,7 +38,20 @@ export function copy<T>(array: Arrayable<T>): Arrayable<T> {
   }
 }
 
+export function splice<T>(array: T[], start: number, k?: number, ...items: T[]): T[]
+export function splice<T>(array: Arrayable<T>, start: number, k?: number, ...items: T[]): Arrayable<T>
+
 export function splice<T>(array: Arrayable<T>, start: number, k?: number, ...items: T[]): Arrayable<T> {
```
Comment (hoxbro) [inline bokehjs/src/lib/core/util/arrayable.ts:44]: I would maybe have named `k` `count`, but that is not related to this PR.

**Review by hoxbro — COMMENTED** [review]
Running http://127.0.0.1:5777/unit/run?k=13831, I get the following console errors (not sure if this is a known or something to look into)

image


I am also curious why `Icons__should_support_all_icons_defined_in_less_icons.less.png` suddenly starts with a right-pointing arrow and is not grouped with the others of its kind.

File: bokehjs/src/lib/core/util/arrayable.ts (hunk 2 above)
Comment (mattpap) [reply bokehjs/src/lib/core/util/arrayable.ts:44]: It should be called `delete_count` to match `splice()`'s original signature.

File: bokehjs/src/lib/core/util/arrayable.ts (hunk 1 above)
Comment (mattpap) [reply bokehjs/src/lib/core/util/arrayable.ts:50]: `toSpliced` is a very new API. It needs to be assessed if we can use it already.

**Comment by mattpap** [comment]
> I get the following console errors (not sure if this is a known or something to look into)

This shouldn't happen. Please report an issue with the relevant details (especially browser version).

**Comment by mattpap** [comment]
> I am also curious why Icons__should_support_all_icons_defined_in_less_icons.less.png suddenly starts with a right-pointing arrow and is not grouped with the others of its kind.

It's as simple grid with no intentional grouping.

File: bokehjs/src/lib/core/util/arrayable.ts (hunk 1 above)
Comment (mattpap) [reply bokehjs/src/lib/core/util/arrayable.ts:50]: It's so new in fact, that TypeScirpt's standard library doesn't have type definitions for it. Not that it's a problem, because we can always extend them with our own.

Comment (hoxbro) [reply bokehjs/src/lib/core/util/arrayable.ts:50]: Looks like it was added in the last two years. My gut feeling is that it is too early to add it.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/toSpliced#browser_compatibility

**Review by hoxbro — APPROVED** [review]
